### PR TITLE
Display schema for the latest 5 reviews inside product's schema

### LINKED ISSUE #23299: Review Schema belongs inside product
10193349-HC

User asked for this issue to be reported here as they do not use GitHub: Issue is that the reviews belong inside the product and are currently not.

This is an example, created with a third  party plugin that has all reviews inside the product, per google requirements. That link shows a good product schema, with all reviews nested inside:

https://search.google.com/structured-data/testing-tool#url=https%3A%2F%2Fdev.japanese-knives.co.il%2Fproduct%2Fshbt_ktsu_gyuto_210_sg2%2F

Now this is the same product, on a different testing server, running normal "Vanilla" Woocommerce schma.

https://search.google.com/structured-data/testing-tool#url=https%3A%2F%2Fstg.japanese-knives.co.il%2Fproduct%2Fshbt_ktsu_gyuto_210_sg2

all of the reviews need to be inside the product, there should not be any independent review schema.

that is what google want: https://developers.google.com/search/docs/data-types/product

They reported that the issue is different than the one being discussed at https://github.com/woocommerce/woocommerce/issues/22969

### PR DESCRIPTION
**Pull request by claudiosanches**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Google recommend to include `review` inside products instead of a new schema just for reviews:

https://developers.google.com/search/docs/data-types/product

I removed the schema for reviews, I kept the method for now, we could deprecate later if necessary.
Note that only loads the latest 5 reviews to avoid performance issues.

By default this Pull Request also fixes the `Either “offers”, “review”, or “aggregateRating” should be specified` error, since it's only one schema for products.

Closes #23299.

### How to test the changes in this Pull Request:

> Create a product and include some reviews on it.
> Check your page (or source code) in https://search.google.com/structured-data/testing-tool/ and check for the errors.
> Now apply the code from this Pull Request and check again on https://search.google.com/structured-data/testing-tool/
> All errors should be fixed (of course should display warnings about brand and global identifier for now).

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Tweak - Include 5 latest reviews to Product's schema.

### CODE REVIEW

File: includes/class-wc-structured-data.php
Comment (rodrigoprimo) [inline includes/class-wc-structured-data.php]: I didn't check, but I believe this is not necessary anymore now that you are using `meta_query`? If this is removed, you might need to change the value passed to `orderby` as well to reference the key in `meta_query`.

File: includes/class-wc-structured-data.php
Comment (rodrigoprimo) [inline includes/class-wc-structured-data.php]: Nice touch to limit the number of reviews to avoid performance issues.

Considering that we are already loading the reviews anyway in the product single page, I wonder if we call `get_comments()` using exactly the same arguments that were used when loading the reviews (or maybe even use a method that calls `get_comments()` internally). This way, when calling it the second time, WP would return a cached response instead of hitting the database twice to get more or less the same data. What do you think of this approach? No worries if it doesn't seem easy to implement or worth it pursuing.

**Review by rodrigoprimo — COMMENTED** [review]
Looks good and works as expected. I just left a few comments with some questions or suggestions.

File: includes/class-wc-structured-data.php
Comment (claudiosanches) [reply includes/class-wc-structured-data.php]: We could wait a little, we had several changes in schema since is created, I can see some change of getting back to that old format, even have both, since display schema individually for all reviews. So it's possible that we may get back to that code. So at least wait a little before try deprecated it.

File: includes/class-wc-structured-data.php
Comment (claudiosanches) [reply includes/class-wc-structured-data.php]: I'll check again, maybe WPCS outdated? But I got a warning for it, that's why I included that notation.

Comment (claudiosanches) [reply includes/class-wc-structured-data.php]: Ohh, you are talking about the `meta_key`? Since we are using `orderby` as `meta_value_num` we should set `meta_key` as well, at least is what WordPress documentation says.

File: includes/class-wc-structured-data.php
Comment (claudiosanches) [reply includes/class-wc-structured-data.php]: I feel the same, and I did look a little at `get_comments()` as well, but we need to skip non-reviews, something that we allow in the front-end, in my changes I also changed the query, so we avoid call any non-review, or reviews with no stars. Because of that I don't think that we could make WP return cached response, or at least should already return for the comment objects in `get_comment_x()` functions.

File: includes/class-wc-structured-data.php
Comment (claudiosanches) [reply includes/class-wc-structured-data.php]: Also having both should at least get merged anyway, but this one is just for sorting.

File: includes/class-wc-structured-data.php
Comment (vedanshujain) [inline includes/class-wc-structured-data.php]: I wonder if we need this comparison. Rating will always be numeric right? Are you using this as a way to check if rating actually exists?

File: includes/class-wc-structured-data.php
Comment (vedanshujain) [inline includes/class-wc-structured-data.php]: comment at line 277 states that we are fetching most recent ratings, but this will actually sort by the value of rating right?

**Review by vedanshujain — COMMENTED** [review]
left some general comments, none of them blocker to PR

File: includes/class-wc-structured-data.php
Comment (claudiosanches) [reply includes/class-wc-structured-data.php]: Yes, checking if is not just some comment without any rating.

File: includes/class-wc-structured-data.php
Comment (claudiosanches) [reply includes/class-wc-structured-data.php]: For sure we don't need it. Just need to last reviews.

**Review by WiredWonder — APPROVED** [review]

**Comment by galapogos01** [comment]
This pull request causes the validation errors from the Review object to move to Product because the itemReviewed object is missing one of offers, review or aggregateRating.  This is due to recent changes in Googles validation of nested sub objects.

**Comment by or128** [comment]
Hello,
All my products now have errors in GSC: Either "offers", "review", or "aggregateRating" should be specified

This is because the reviews need to be nested inside the product data.

**Comment by claudiosanches** [comment]
Pull Request update, all errors fixed:

Screenshot from 2019-10-22 18-03-49

I got 4 warnings because I'm in a test environment:

Screenshot from 2019-10-22 18-03-58

The another two are about brand and identifier:

Screenshot from 2019-10-22 18-04-08

**Comment by or128** [comment]
> Pull Request update, all errors fixed:

Was that a suggestion I should follow? What do you mean? will that solve the incorrect way the schema is outputted?

**Comment by drnick2** [comment]
So what is the solution to this problem?  I currently have 36 new errors in Google Search Console and the list is growing each day.

**Comment by claudiosanches** [comment]
@TimTank the code in this Pull Request is the fix.

**Comment by TimTank** [comment]
> @TimTank the code in this Pull Request is the fix.

Srry for waisting your time, I'm a newbie. Could you explain what to do, one step at the time?

**Comment by zigojacko** [comment]
@claudiosanches I think you're trying to fix a problem that is different to the one recently came about (October 2019) and reported by @or128 above which has only come about now due to a change by Google in how they read Product Schema.

This new issue, whilst the same error, is specifically related to individual product review Schema within the Product Schema type.

The latest pull-request or version of WooCommerce does not resolve this issue. This issue only started happening to thousands of WooCommerce stores from 16th October as per [this update by Google](https://support.google.com/webmasters/answer/6211453#rich_result_reports).

**Comment by claudiosanches** [comment]
@zigojacko I think you missed this comment: https://github.com/woocommerce/woocommerce/pull/24278#issuecomment-545155362

I'm fixing all errors on the schema, also removing the independent Review Schema and including all into the Product Schema, so there's no need to duplicated any data in more than one schema.

@TimTank So it's better wait for the next release, since it may require some dev skills.

**Comment by kschmidtdev** [comment]
I realize this PR is just for addressing the review schema, but all the related issues have been closed. What if you have reviews disabled on your woocommerce store? Will there be a fix to include the newly-required "offers" field?

**Comment by galapogos01** [comment]
I can confirm this PR fixes all errors in both scenarios - products with a review, and products without a review.
Products without a review still generate a warning as expected because they do not have an aggregateRating.

Thanks @claudiosanches for re-cutting this PR.  I hope it gets into base quickly because this issue impacts SEO quite seriously.

Cheers,
Jason

**Comment by claudiosanches** [comment]
@kschmidtdev there's already support for `offers` since the first time we implemented this feature.

> but all the related issues have been closed. 

Seems like you didn't followed it closely, most of the errors is because the independent reviews schema, since Google now requires all fields to be included into the Reviews Schema, this PR remove this independent Review Schema, and included reviews inside the Products Schema, so no need to duplicate any data like before.

@galapogos01 thanks for testing and giving feedback.

**Comment by or128** [comment]
Version 3.8 was just released, and still schema is not fixed.

**Why?**

**Comment by stephenblender** [comment]
@or128 - I temporarily added this file "includes/class-wc-structured-data.php" to my existing wc plugin which worked fine and corrected issues. Hopy very temporary while waiting for this update to find it's way into the release.

**Comment by drnick2** [comment]
@stephenblender Can you elaborate or explain what you actually replaced and with what file? Thanks!

**Comment by stephenblender** [comment]
The file I mentioned is one of (the?) file that has been pulled and updated here, and is waiting to be approved to be merged back into a future release of woocommerce. Rather than wait I replaced this same file in my current wc plugin with the updated one. I'm not the best person to seek advice from on this as I am a bit of a hack...just offering some ideas for the desperate. If this is unclear it might be something you do not want to mess with. This 'fix' is certainly not best practice.

**Comment by drnick2** [comment]
@stephenblender  Thanks for clearing that up and explaining what you did, I appreciate it.

**Comment by or128** [comment]
Hello,
Can someone give a step by step guide on how to fix the issue?
I think waiting for a fix could take months.

**Comment by or128** [comment]
> @or128 - I temporarily replaced this file "includes/class-wc-structured-data.php" to my existing wc plugin which worked fine and corrected issues. Hopefully very temporary while waiting for this update to find it's way into the release.

@stephenblender I don't understand what you did with the file. I can locate it just in the includes folder.

**Comment by or128** [comment]
How can I download the new version of that file? I can't seem to find anyway to get it to my PC.

**Comment by or128** [comment]
OK, so I think I've managed to get the revised file from here: https://github.com/woocommerce/woocommerce/blob/3c92523d9272dd80be7f8d9cb09cf487bf94e475/includes/class-wc-structured-data.php

It does not solve the issue.
The reviews did go into the product, that was good, but this error persists:
_One of offers or review or aggregateRating should be provided._

I think Google changed something fundamental in the review schema that is not addressed by this fix. This fix only puts the reviews into the product (which is a good thing).

**Review by or128 — CHANGES_REQUESTED** [review]
OK, so I think I've managed to get the revised file from here: https://github.com/woocommerce/woocommerce/blob/3c92523d9272dd80be7f8d9cb09cf487bf94e475/includes/class-wc-structured-data.php

It does not solve the issue.
The reviews did go into the product, that was good, but this error persists:
_One of offers or review or aggregateRating should be provided._

I think Google changed something fundamental in the review schema that is not addressed by this fix. This fix only puts the reviews into the product (which is a good thing).

**This error should be resolved: _One of offers or review or aggregateRating should be provided._**

**Comment by claudiosanches** [comment]
@or128 sounds like you didn't managed to proper test it, since the errors are the same as before.

**Comment by or128** [comment]
> @or128 sounds like you didn't managed to proper test it, since the errors are the same as before.

Yes, that's exactly what I've said.

**Comment by claudiosanches** [comment]
@or128 still even that our team is not able yet to review it this is the fix for nesting (and also remove the schema for reviews since is already merged into products schema). 
In other words, this Pull Request fix both problems reported by you, so please, if you have any additional information that may help us with this fixes please post in this Pull Request. Thank you for your collaboration.

**Comment by or128** [comment]
@claudiosanches

I do have additional information - this fix only nests the reviews into the product schema, but that does not solve the error.

**Comment by or128** [comment]
@claudiosanches

Here you can see that this latest fixed _class-wc-structured-data.php_ does not solve the error:
https://search.google.com/structured-data/testing-tool#url=https%3A%2F%2Ftst.japanese-knives.co.il%2Fproduct%2Fymw_santoku_vg10_180_wh%2F

This is on a test server where I've switched the _class-wc-structured-data.php_ file for your one.

This is the product: https://tst.japanese-knives.co.il/product/ymw_santoku_vg10_180_wh/

**Comment by stephenblender** [comment]
The problem looks like the nested reviews themselves have product references "itemReviewed" which google identifies as incomplete. Note the errors are not for the product itself but for the nested products within the reviews (within the product).

This is not happening on my website so I wonder if your theme or another extension is adding product references to the reviews. Test with all other plugins turned off and stock theme. if technical look for code filtering "woocommerce_structured_data_review".

**Comment by juliaamosova** [comment]
@claudiosanches there is another PR to fix this issue: https://github.com/woocommerce/woocommerce/pull/25003.

**Comment by khag7** [comment]
@stephenblender is correct. Each reviews is now required to provide full product data, even if the product data is already provided elsewhere. If you have a product with 9 reviews, each review has to as part of it's data provide the product info, resulting in the product info being sent 10 separate times in the same structured data set. Yes it's absurd but short of removing reviews completely, my PR is a reasonable solution.

**Comment by stephenblender** [comment]
To clarify: using google structure data testing tool it appears reviews nested within a product schema do not require that they have nested product within them ('itemreviewed'). Distinct reviews schema that contain 'itemreviewed' do appear to require more detailed information. 

https://search.google.com/structured-data/testing-tool/u/0/#url=https%3A%2F%2Fvitaminb12direct.com%2Fproduct%2Finjectable-vitamin-b12-1000mcg%2F

**Comment by khag7** [comment]
If that's the case then the PR here is better than mine as long as reviews in this PR are nested in the product rather than listed separately.

**Comment by or128** [comment]
> To clarify: using google structure data testing tool it appears reviews nested within a product schema do not require that they have nested product within them ('itemreviewed'). Distinct reviews schema that contain 'itemreviewed' do appear to require more detailed information.
> 
> https://search.google.com/structured-data/testing-tool/u/0/#url=https%3A%2F%2Fvitaminb12direct.com%2Fproduct%2Finjectable-vitamin-b12-1000mcg%2F

I 100% agree.
The _itemReviewed_ Schema is to be used if you have, for example, a webpage that is a review for lets say a TV, and then the main Schema would be review, and nested inside it is a reference to the reviewed product using the _itemReviewed_.

when the main subject of the page is the product, and the reviews are already nested inside it, there should be no _itemReviewed_.

**Comment by or128** [comment]
Also, why is the body of the review not outputted?

**Comment by or128** [comment]
OK, I offer my code.
I believe it solves 100% of the issues:

`if ( $comments ) {
				$markup['review'] = array();
				foreach ( $comments as $comment ) {
					$markup['review'][] = array(
						'@type'         => 'Review',
						'author'        => array(
							'@type' => 'Person',
							'name'  => get_comment_author( $comment->comment_ID ),
						),
						'reviewBody' => $comment->comment_content,
						'reviewRating'  => array(
							'@type'       => 'Rating',
							'ratingValue' => get_comment_meta( $comment->comment_ID, 'rating', true ),
						),
						'datePublished' => get_comment_date( 'c', $comment->comment_ID ),
					);
				}
			}
		}`

**Comment by claudiosanches** [comment]
@juliaamosova that PR it's not necessary once this one is merged, since we are nesting the Review's schema inside the Product's schema.

**Comment by claudiosanches** [comment]
@or128 become very confuse to follow what you are saying, there's several different things you are commenting on here, makes very hard to take any action as well.
Can you please summarize things? Also every time you doing changes on your store and sending us a link from Google doesn't help, since the results will change when I try the link.

**Comment by zigojacko** [comment]
> @or128 become very confuse to follow what you are saying, there's several different things you are commenting on here, makes very hard to take any action as well.
> Can you please summarize things? Also every time you doing changes on your store and sending us a link from Google doesn't help, since the results will change when I try the link.

@claudiosanches it has been explained by numerous people commenting multiple times on both this thread and the other(s). It seems like it is you that isn't grasping the issue here because as I explained many comments ago now, your PR does not fix the issue originally reported. (I've been following the comments ever since but gave up trying to explain the problem).

If you look at @or128's last comment, it displays his test site in Google's structured data tool and with changes to the necessary file within the WooCommerce plugin along with screenshots of both the Structured Data Tool and codebase.

It clearly shows correct without the error that we were all reporting countless times a while ago in these comments. This is exactly what is needed to be done. I'm not sure what you are getting confused about...? Thanks.

**Comment by claudiosanches** [comment]
@zigojacko please check that the last code from @or128 it's exactly the same as my code in this PR, but he introduced a `reviewBody`, what is not required by Google, not even the problem that this PR is trying to fix.
Have you tried to use the code from this PR?

**Comment by or128** [comment]
@claudiosanches it is not the same as your code.
I've removed the _itemReviewed_, since it is the trigger of the erros.

**Comment by claudiosanches** [comment]
@or128 the code you posted in https://gist.github.com/or128/ac553974d69c66dda76f649eed88ab27 it's the same code as my code in this PR, this is the only difference:

Screenshot from 2019-11-12 11-57-58

>  I've removed the itemReviewed, since it is the trigger of the erros.

Also you can check the code in this PR, there's no more `itemReviewed` for a long time.

**Comment by or128** [comment]
@claudiosanches Here's a summery for you.

**Current Issues:**
1. Schema error: Either “offers”, “review”, or “aggregateRating” should be specified
2. Reviews should be nested within the product Schema (that was indeed solved by @claudiosanches)
3. Review body is not outputted.

**Solutions:**
1. Removing the _itemReviewed_ from the review schema (it shouldn't be outputted).
2. Just toke @claudiosanches code, that was good.
3. Added the review body to the Schema: `'reviewBody' => $comment->comment_content,`

**Comment by claudiosanches** [comment]
@or128 thank you.

Let's talk about each issue:

> Schema error: Either “offers”, “review”, or “aggregateRating” should be specified

This error comes from the Review's schema, since this schema isn't nested in the Product's schema this requires to have `itemReviewed` including “offers”, “review”, or “aggregateRating”.

This error should not happens anymore, since I removed completely the Review's schema from displaying in the product's page: https://github.com/woocommerce/woocommerce/pull/24278/files#diff-4c5df99bf6db5658791ac705bea97f8eL32

>  Reviews should be nested within the product Schema (that was indeed solved by @claudiosanches)

I did it to solve the first issue about missing “offers”, “review”, or “aggregateRating”, those are already included into each product, and since there's already a Product's schema in the page I nested the reviews as well.

At this point we don't have more a "Review's schema" in the product page, only a "Product's Schema" including all product data and also nesting all review data inside.
So both issues are fixed.

> Review body is not outputted.

This is a new issue, that you are raising now, and it's not the subject of this change, not even in the old Review Schema we include it, Google also doesn't require it, that's why is not previously included. But we can include it now, no problem of doing suggestions like that.

**Comment by or128** [comment]
@claudiosanches I truly don't know what to say anymore, everything was explained repeatedly by me and others, I don't know what explanation would make you understand.

The code I offer is a complete solution, it should be released after tests.

**Comment by claudiosanches** [comment]
@or128 You don't agree that my code fix the two issues you listed? On your code you only include `reviewBody`, this is the only difference from your code and my code in this PR.

**Comment by or128** [comment]
> @or128 You don't agree that my code fix the two issues you listed? On your code you only include `reviewBody`, this is the only difference from your code and my code in this PR.

When I've tried it, it did not solve the error.
Maybe you've changed it since then.

Anyhow, just release my code, you can take the credit, but my code is correct.

**Comment by claudiosanches** [comment]
@or128 my last change on this code is from `Oct 22, 2019`, your gist is from 8 hours ago. I'm not talking about credit, I'm confuse because you said that the code on this Pull Request doesn't work at all for you!

**Comment by or128** [comment]
I've never used github before, I'm just a shop owner.
I don't even know what a pull request is.

**Comment by claudiosanches** [comment]
@or128 ok, no problem, just say that you are not able to proper test this Pull Request and will be fine, no reason to generate more confusion.
At least from your snippet that is the same code that this Pull Request you have been able to confirm that my code works. So thanks.
I'll check later for `reviewBody` in a new Pull Request, thanks for suggesting including it.

**Comment by zigojacko** [comment]
@claudiosanches I on the other hand did try the modified file that was included in your first disclosed PR and it didn't fix the issue, that was several weeks ago now. I think there is confusion stemmed from the fact that there is multiple comment threads for the same issue - I can't even find my original comment (although I think in each one, you linked to this same PR). Not sure if it was modified after the first release though or something. I have not tried updating it on a live site since I did back then when I commented but we know that the code above from @or128 definitely resolves the issue and if (however it happened) this PR is now exactly the same, then we just need someone to test it on a live site and see if it does the trick (I'm unable to right now as busy with some other projects).

**Comment by claudiosanches** [comment]
@zigojacko 

> I have not tried updating it on a live site since I did back then when I commented but we know that the code above from @or128 definitely resolves the issue and if (however it happened) this PR is now exactly the same

21 days ago I made the change that fixed all issues you all are reporting and also I included screenshots: https://github.com/woocommerce/woocommerce/pull/24278#issuecomment-545155362

You have commented 3 days after it claiming that code on this PR doesn't fix anything: https://github.com/woocommerce/woocommerce/pull/24278#issuecomment-546256900

Everything that you clam in this comment is already fixed before you commented...
After that instead of you trying to test my code, you return here trying to offend me, and dismiss all my hard work to fix this issue:

> @claudiosanches it has been explained by numerous people commenting multiple times on both this thread and the other(s). It seems like it is you that isn't grasping the issue here because as I explained many comments ago now, your PR does not fix the issue originally reported. (I've been following the comments ever since but gave up trying to explain the problem).

This doesn't sound polite, this really offend me, I only want to fix this issue, but everyone keeps giving me incorrect feedback without even testing this code.

**Comment by stephenblender** [comment]
For the record, my experience and understanding:
- this update fixed the errors on my website. I linked to google structured data tester of my page earlier in the comments. No errors in tester or in search console.

- I think it is odd the included reviews (nested) do not include review body, BUT google's example of a product the nested review does not include a review body

- my products show no errors, but my 'search appearance' for the product page has not reverted to product yet...a concern. no stars, no price in search engine result page. This could just be lag in updating google index I suppose. 

- I wonder about including only 5 reviews, but have no experience regarding the server impact. I could not find a reference in google product discussion on ideal number to include.

**Comment by claudiosanches** [comment]
@stephenblender thank you for your comment, and I'm very glad that someone is taking time to responsibly test it.

> I think it is odd the included reviews (nested) do not include review body, BUT google's example of a product the nested review does not include a review body

We haven't included review body not even before with the independent Review Schema, that's why is not included in this change right now.
But this is something easy to include later if necessary, I'm planing a new Pull Request after this is merged to include it.

> my products show no errors, but my 'search appearance' for the product page has not reverted to product yet...a concern. no stars, no price in search engine result page. This could just be lag in updating google index I suppose.

Google takes some time to update it.

> I wonder about including only 5 reviews, but have no experience regarding the server impact. I could not find a reference in google product discussion on ideal number to include.

A product can have infinite number of reviews, but Google seems to not require to load all of those, also with `aggregateRating` have rating already listed, so just need some reviews to display on Google. But this may be subject of change later, also not hard to increase the number of reviews if necessary.

Thanks again for your help!

**Comment by or128** [comment]
> @claudiosanches we just need someone to test it on a live site and see if it does the trick (I'm unable to right now as busy with some other projects).

I am running it (my code) on my live store: https://www.japanese-knives.co.il/

**Comment by or128** [comment]
@claudiosanches What is the correct way of getting the latest version of your code from this thread?
it seems people are downloading and testing something you did not intend them to download.

**Comment by or128** [comment]
> A product can have infinite number of reviews, but Google seems to not require to load all of those, also with `aggregateRating` have rating already listed, so just need some reviews to display on Google. But this may be subject of change later, also not hard to increase the number of reviews if necessary.

Yes, 5 is enough.

**Comment by claudiosanches** [comment]
@or128 you can download from https://github.com/woocommerce/woocommerce/archive/fix/23299.zip

**Comment by peterfabian** [comment]
I've tested this myself, with and without reviews and the errors on the Google's testing site are gone. 

I think part of the confusion comes from the fact that the code @or128 tested (the file linked initially in the comments) is not the latest version from this PR, it is just the first commit. 
This is the latest version of the file: https://github.com/woocommerce/woocommerce/blob/237a047e26a1adcf649b5596bfd473e835dc862c/includes/class-wc-structured-data.php, thus it was still showing errors.

@zigojacko please let's try to keep things respectful here. Reading through the backscroll, it seems to me that moving the reviews into the product schema should produce valid result with no errors. Please let us know if we're missing something, ideally with steps how to create a product which manifests the problem.

Now to the changes:
1. when the reviews are **turned off**, the product schema is valid, as it contains `offers` section

2. when the reviews are **turned on**, 
  a) if the product has no reviews, there is no review section
  b) if the product has reviews, the average rating and reviews are included:
```
"aggregateRating": {
        "@type": "AggregateRating",
        "ratingValue": "3.50",
        "reviewCount": 2
      },
      "review": [
        {
          "@type": "Review",
          "author": { "@type": "Person", "name": "demo" },
          "reviewRating": { "@type": "Rating", "ratingValue": "3" },
          "datePublished": "2019-11-12T17:09:31+00:00"
        },
        {
          "@type": "Review",
          "author": { "@type": "Person", "name": "demo" },
          "reviewRating": { "@type": "Rating", "ratingValue": "4" },
          "datePublished": "2019-11-12T17:09:16+00:00"
        }
      ]
```

In all these cases, Google testing tool reports no errors for me and the preview it produces seems to be valid:
Screenshot 2019-11-12 at 19 00 37

I'm inclined to approve this change as it seems to fix the problem, unless I would hear from you folks with instructions how to set up a product that shows the problem.

**Comment by or128** [comment]
> @or128 you can download from https://github.com/woocommerce/woocommerce/archive/fix/23299.zip

How was I suppose to find it without asking you?

**Comment by peterfabian** [comment]
@or128 
Apologies, GitHub is a forum primarily aimed for developers. We, unfortunately, don't have much capacity to explain how to test PRs and fixes. Especially more elaborate fixes with multitude of files are more difficult to test, more so now that we have a build step in creating the WooCommerce package. Therefore it's better suited for testing by advanced users, other developers and dev agencies.  
Any help is appreciated, but as you can see in this PR, sometimes it can create more confusion. 

> How was I suppose to find it without asking you?

This is something that's part of 'standard' interface on GitHub. Any branch on a dev project is available via 
https://github.com/org_name/repo_name/archive/branch_name.zip

Other way how to get the code to your server is e.g. by applying a patch file, which you can get by visiting Pull Request's URL and appending `.patch` at the end, e.g. for this one:
https://github.com/woocommerce/woocommerce/pull/24278.patch

It's apparent that we should write a guide about how to test a Pull Request change on a shop for merchants less familiar with development. We would generally recommend getting involved by testing beta versions and release candidates on your testing/staging copy website with [Beta tester plugin](https://wordpress.org/plugins/woocommerce-beta-tester/), but I understand you might want to try out the fixes sooner, which is great and thanks for actively trying to help!

**Comment by dylande2ker** [comment]
We are really struggling with this problem, for all our customers with a WooCommerce webshop, hoping for a solution.

**Comment by or128** [comment]
Please add the review body too.

**Review by peterfabian — APPROVED** [review]
Thanks for adding the `reviewBody`.

I've tested on a product with 0, 1 or multiple reviews, it works fine.
Same when the reviews are turned off.

**Comment by stephenblender** [comment]
Just a note that my site which has this installed (without review body) has ranked again for the past two days in search console performance metrics "rich results" for product. For the past month it had dropped out of rich results and was showing in search only as a web page.

Actually ranking better...maybe because other sites have not yet returned to the mix? I'll take it. Nice work.

**Comment by mrkapable007** [comment]
Hi, I have over 2500 errors on my search console for this "review issue" , can someone tell me the code to fix this?

**Comment by or128** [comment]
Is this included in 3.8.1?

**Comment by peterfabian** [comment]
@or128 Yes, it is. My testing site is showing no errors, have you had a chance to test?

**Comment by away1112** [comment]
The error still on the shop base page,anyone have same problem ? Thanks

**Comment by peterfabian** [comment]
@away1112 Can you please share the error you see?

**Comment by TimTank** [comment]
I still have the same errors as before. Just crawld and checked again.

Missing field review, aggregaterating, brand and missing product ID.

**Comment by away1112** [comment]
> @away1112 Can you please share the error you see?

I am sorry, just SEO plugin not upgrade for 3.8.1,so disable plugin Snippet Data on shop base page showing no errors,Thank you very much.

**Comment by peterfabian** [comment]
@TimTank 

As far as I understand, those are _warnings_, not errors. 
This PR got rid of the _errors_ you would get on the reviews.

- fields `review` and `aggregateRating` are included if the product has reviews. If it doesn't we can't really provide any value in there (Is the aggregatRating 0 or 5 or 2.5 if the product doesn't have any reviews?)
- `brand` is currently not supported in WC core, you can add it by using [WooCommerce Brands extension](https://woocommerce.com/products/brands/)
- similarly with `GTIN8`. We use SKU at the moment, but are considering adding GTIN in the future.

**Comment by bondimedical3** [comment]
I submitted a validation in Google Search Console for the latest fix and it was successful this morning.

**Comment by drnick2** [comment]
My validation just came back positive, so thanks to all who contributed and worked on the fix for this issue.

**Comment by TimTank** [comment]
> @TimTank
> 
> As far as I understand, those are _warnings_, not errors.
> This PR got rid of the _errors_ you would get on the reviews.
> 
> * fields `review` and `aggregateRating` are included if the product has reviews. If it doesn't we can't really provide any value in there (Is the aggregatRating 0 or 5 or 2.5 if the product doesn't have any reviews?)
> * `brand` is currently not supported in WC core, you can add it by using [WooCommerce Brands extension](https://woocommerce.com/products/brands/)
> * similarly with `GTIN8`. We use SKU at the moment but are considering adding GTIN in the future.
> 
> Example report from testing site running WC 3.8.1, product with reviews:
> Screenshot 2019-11-29 at 12 09 34
> 
> product with no reviews:
> Screenshot 2019-11-29 at 12 08 16

Just checked, you're completely right! 

Thank you guys for all your great work!

**Comment by Webber2014** [comment]
In the meanwhile something has changed with Google and the warnings, are now seen as errors.
You also get notified that there are errors.

I have the latest versions of WordPress and WooCommerce installed, tried all solutions in this and other threads (even modifying and adding other code), disabled all plugins, tried (WC) markup scheme plugins, tried other themes but nothing solves these now by Google considered errors.

Do see some having apparently solved the problem but with every Woocommerce webshop I manage I get the same errors that review and aggregateRating are missing. 
I do have product reviews disabled in several webshops - but those are not the reviews meant by Google I suspect?

Any suggestions?

**Comment by peterfabian** [comment]
Hi @Webber2014 , if you see an error that hasn't been reported yet (i.e. is different from https://github.com/woocommerce/woocommerce/issues/25666), please open a new issue in the repository.

**Comment by Webber2014** [comment]
Hi @peterfabian - I am quite sure it is the same issue as described here.

**Comment by stephenblender** [comment]
Although your symptom may be the same  as this thread I do not think it is the same problem. This issue was corrected with changes to wc that have gone live. If you are using the latest wc this thread should not fix your problem. 
Unfortunately if your symptom is the same then all the discussion of this problem will make your research harder. I suggest starting with a test site, stock theme, wc, and one at a time adding complexity until you  isolate the cause.

**Comment by Webber2014** [comment]
Hi @stephenblender - I do believe I am not alone, found several similar threads here and there with ppl stating having the same issue who can't solve it either. I personally also have it with multiple webshops. A though cookie ... :-(

**Comment by peterfabian** [comment]
@Webber2014 please create a new issue and include relevant information about your setup from [system status report](https://docs.woocommerce.com/document/how-to-review-your-woocommerce-system-status/), your review settings and product setup so that we can reproduce the issue and we'll try to look into it soon.

**Pull request merged**
